pkg/provider/beyondtrust/provider.go

```diff
@@ -0,0 +1,327 @@
+	}
+
+	// validate inputs
+	errorsInInputs := utils.ValidateInputs(clientID, clientSecret, &apiURL, clientTimeOutInSeconds, &separator, config.VerifyCA, logger, certificate, certificateKey, &retryMaxElapsedTimeMinutes, &maxFileSecretSizeBytes)
+
+	if errorsInInputs != nil {
+		return nil, fmt.Errorf("error: %w", errorsInInputs)
+	}
+
+	// creating a http client
+	httpClientObj, err := utils.GetHttpClient(clientTimeOutInSeconds, config.VerifyCA, certificate, certificateKey, logger)
+
+	if err != nil {
+		return nil, fmt.Errorf("error: %w", err)
+	}
+
+	// instantiating authenticate obj, injecting httpClient object
+	authenticate, _ := auth.Authenticate(*httpClientObj, backoffDefinition, apiURL, clientID, clientSecret, logger, retryMaxElapsedTimeMinutes)
+
+	return &Provider{
+		apiURL:        config.APIURL,
+		retrievaltype: config.Retrievaltype,
+		authenticate:  *authenticate,
+		log:           *logger,
+		separator:     separator,
+	}, nil
+}
+
+func loadConfigSecret(ctx context.Context, ref *esv1beta1.BeyondTrustProviderSecretRef, kube client.Client, defaultNamespace string) (string, error) {
+	if ref.SecretRef == nil {
+		return ref.Value, nil
+	}
+
+	if err := validateSecretRef(ref); err != nil {
+		return "", err
+	}
+
+	namespace := defaultNamespace
+	if ref.SecretRef.Namespace != nil {
+		namespace = *ref.SecretRef.Namespace
+	}
+
+	ESOLogger.Info("using k8s secret", "name:", ref.SecretRef.Name, "namespace:", namespace)
+	objKey := client.ObjectKey{Namespace: namespace, Name: ref.SecretRef.Name}
+	secret := v1.Secret{}
+	err := kube.Get(ctx, objKey, &secret)
+	if err != nil {
+		return "", err
+	}
+
+	value, ok := secret.Data[ref.SecretRef.Key]
+	if !ok {
+		return "", fmt.Errorf(errNoSuchKeyFmt, ref.SecretRef.Key)
+	}
+
+	return string(value), nil
+}
+
+func validateSecretRef(ref *esv1beta1.BeyondTrustProviderSecretRef) error {
+	if ref.SecretRef != nil {
+		if ref.Value != "" {
+			return errSecretRefAndValueConflict
+		}
+		if ref.SecretRef.Name == "" {
+			return errMissingSecretName
+		}
+		if ref.SecretRef.Key == "" {
+			return errMissingSecretKey
+		}
+	} else if ref.Value == "" {
+		return errSecretRefAndValueMissing
+	}
+
+	return nil
+}
+
+func (p *Provider) GetAllSecrets(_ context.Context, _ esv1beta1.ExternalSecretFind) (map[string][]byte, error) {
+	return nil, fmt.Errorf("GetAllSecrets not implemented")
+}
+
+// GetSecret reads the secret from the Password Safe server and returns it. The controller uses the value here to
+// create the Kubernetes secret.
+func (p *Provider) GetSecret(_ context.Context, ref esv1beta1.ExternalSecretDataRemoteRef) ([]byte, error) {
+	managedAccountType := !strings.EqualFold(p.retrievaltype, "SECRET")
+
+	retrievalPaths := utils.ValidatePaths([]string{ref.Key}, managedAccountType, p.separator, &p.log)
+
+	if len(retrievalPaths) != 1 {
+		return nil, fmt.Errorf(errInvalidRetrievalPath)
+	}
+
+	retrievalPath := retrievalPaths[0]
+
+	_, err := p.authenticate.GetPasswordSafeAuthentication()
```

Comment:
What is the purpose of this call? 🤔  Because we don't care about the return value and the token it generates and signed isn't stored anywhere in the auth struct.

Reply:
Yes, it creates a session internally in our API to be able to call the rest of the endpoints